File: 추천50제/완전탐색.백트래킹/2798.py
import sys
import itertools

input = sys.stdin.readline

n,m = map(int,input().split())
arr = list(map(int,input().split()))
result = 0
combinations = list(itertools.combinations(arr,3))
for i in combinations:
    if sum(i)>m:
        continue
    result = max(result,sum(i))
print(result)

File: BOJ/2798.js
let fs = require('fs');
let dev = true;
let input = fs
  .readFileSync(dev ? './test.txt' : './dev/stdin')
  .toString()
  .trim()
  .split('\n');

let [n, m] = input.shift().split(' ').map(Number);
let arr = input.shift().split(' ').map(Number);

const makeCombination = function (arr, n) {
  let results = [];
  if (n == 1) return arr.map((e) => [e]);
  arr.forEach((value, index, origin) => {
    const rest = origin.slice(index + 1);
    const combi = makeCombination(rest, n - 1);
    const attached = combi.map((e) => [value, ...e]);
    results.push(...attached);
  });
  return results;
};
console.log(
  Math.max(
    ...makeCombination(arr, 3)
      .map((e) => {
        return e.reduce((acc, cur) => acc + cur);
      })
      .filter((e) => e <= m)
  )
);
